src/framerate_gui.cpp

```diff
@@ -0,0 +1,752 @@
+			this->num_valid = min(NUM_FRAMERATE_POINTS, this->num_valid + 1);
+		}
+
+		void AddAccumulate(TimingMeasurement duration)
+		{
+			this->durations[this->prev_index] += duration;
+		}
+
+		void AddPause(TimingMeasurement start_time)
+		{
+			if (this->durations[this->prev_index] != INVALID_DURATION) {
+				this->timestamps[this->next_index] = start_time;
+				this->durations[this->next_index] = INVALID_DURATION;
+				this->prev_index = this->next_index;
+				this->next_index += 1;
+				if (this->next_index >= NUM_FRAMERATE_POINTS) this->next_index = 0;
+				this->num_valid += 1;
+			}
+		}
+
+		double GetAverageDurationMilliseconds(int count)
+		{
+			count = min(count, this->num_valid);
+
+			int first_point = this->prev_index - count;
+			if (first_point < 0) first_point += NUM_FRAMERATE_POINTS;
+
+			double sumtime = 0;
+			for (int i = first_point; i < first_point + count; i++) {
+				auto d = this->durations[i % NUM_FRAMERATE_POINTS];
+				if (d != INVALID_DURATION) {
+					sumtime += d;
+				} else {
+					count--;
+				}
+			}
+
+			if (count > 0) {
+				return sumtime * 1000 / count / TIMESTAMP_PRECISION;
+			} else {
+				return 0;
+			}
+		}
+
+		/** Get current rate of a performance element, based on approximately the past one second of data */
+		double GetRate()
+		{
+			int point = this->prev_index;
+			int last_point = this->next_index - this->num_valid;
+			if (last_point < 0) last_point += NUM_FRAMERATE_POINTS;
+
+			int count = 0;
+			auto last = this->timestamps[point];
+			TimingMeasurement total = 0;
+
+			while (point != last_point) {
+				if (this->durations[point] != INVALID_DURATION) {
+					total += last - this->timestamps[point];
+				}
+				last = this->timestamps[point];
+				if (total >= TIMESTAMP_PRECISION) break;
+				point--;
+				count++;
+				if (point < 0) point = NUM_FRAMERATE_POINTS - 1;
+			}
+
+			if (total > 0 && count > 0) {
+				return (double)count * TIMESTAMP_PRECISION / total;
+			} else {
+				return 0;
+			}
+		}
+	};
+
+	static const double GL_RATE = 1000.0 / MILLISECONDS_PER_TICK;
+
+	PerformanceData _pf_data[PFE_MAX] = {
+		PerformanceData(GL_RATE),               // PFE_GAMELOOP
+		PerformanceData(1),                     // PFE_ACC_GL_ECONOMY
+		PerformanceData(1),                     // PFE_ACC_GL_TRAINS
+		PerformanceData(1),                     // PFE_ACC_GL_ROADVEHS
+		PerformanceData(1),                     // PFE_ACC_GL_SHIPS
+		PerformanceData(1),                     // PFE_ACC_GL_AIRCRAFT
+		PerformanceData(GL_RATE),               // PFE_DRAWING
+		PerformanceData(1),                     // PFE_ACC_DRAWWORLD
+		PerformanceData(60.0),                  // PFE_VIDEO
+		PerformanceData(1000.0 * 8192 / 44100), // PFE_SOUND
+	};
+
+}
+
+
+static std::chrono::high_resolution_clock _hr_clock;
+static TimingMeasurement GetPerformanceTimer()
+{
+	auto usec = std::chrono::time_point_cast<std::chrono::microseconds>(_hr_clock.now());
+	return usec.time_since_epoch().count();
+}
```

Comment:
_hr_clock is only used once. Can you not just call std::chrono::high_resolution_clock.now() directly?

Reply:
Some implementations might have state kept around in the clock object, in fact I had hoped Microsoft's had had that. (It calls QueryPerformanceFrequency every call to `now()` even though the frequency should never change while the system is running.) Of course it could also be a static variable inside the function.

Reply:
Wait, didn't realize now() was a static function... huh.